reference/7.1/Microsoft.PowerShell.Core/About/about_Functions_Advanced_Argument_Completion.md

```diff
@@ -0,0 +1,367 @@
+options. the specified values are available
+
+The syntax is as follows:
+
+```powershell
+function Test-ArgumentCompletions {
+    [CmdletBinding()]
+    param (
+        [Parameter(Mandatory=$true)]
+        [ArgumentCompletions('Fruits', 'Vegetables')]
+        $Type,
+
+        [Parameter()]
+        [ArgumentCompletions('Apple', 'Banana', 'Orange')]
+        $Fruit,
+
+        [Parameter()]
+        [ArgumentCompletions('Tomato', 'Corn', 'Squash')]
+        $Vegetable
+    )
+}
+```
+
+Each of the parameters is provided a list of options to the
+**ArgumentCompletions** attribute to enable tab completion.
+
+This attribute was introduced in PowerShell Core 6.0
+
+## Class-based argument completers
+
+Beginning in PowerShell 7.2, a new feature was added that allows you to define
+more generic implementations of parameterized argument completers.
+
+By deriving from `ArgumentCompleterAttribute`, it's possible to create generic
+completers that can be reused, for example:
+
+```powershell
+[DirectoryCompleter(ContainingFile="pswh.exe", Depth=2)]
+
+[DateCompleter(WeekDay='Monday', From="LastYear")]
+
+[GitCommits(Branch='release')]
+```
+
+The derived attributes must implement the `IArgumentCompleterFactory` interface
+and use property values to create a specialized completer.
+
+```powershell
+using namespace System.Collections
+using namespace System.Collections.Generic
+using namespace System.Management.Automation
+using namespace System.Management.Automation.Language
+
+class NumberCompleter : IArgumentCompleter {
+
+    [int] $From
+    [int] $To
+    [int] $Step
+
+    NumberCompleter([int] $from, [int] $to, [int] $step) {
+        if ($from -gt $to) {
+            throw [ArgumentOutOfRangeException]::new("from")
+        }
+        $this.From = $from
+        $this.To = $to
+        $this.Step = $step -lt 1 ? 1 : $step
+    }
+
+    [IEnumerable[CompletionResult]] CompleteArgument(
+        [string] $CommandName,
+        [string] $parameterName,
+        [string] $wordToComplete,
+        [CommandAst] $commandAst,
+        [IDictionary] $fakeBoundParameters) {
+
+        $resultList = [List[CompletionResult]]::new()
+        $local:to = $this.To
+        $local:step = $this.Step
+        for ($i = $this.From; $i -lt $to; $i += $step) {
+            $resultList.Add([CompletionResult]::new($i.ToString()))
+        }
+
+        return $resultList
+    }
+}
+
+class NumberCompletionsAttribute : ArgumentCompleterAttribute, IArgumentCompleterFactory {
+    [int] $From
+    [int] $To
+    [int] $Step
+
+    NumberCompletionsAttribute([int] $from, [int] $to, [int] $step) {
+        $this.From = $from
+        $this.To = $to
+        $this.Step = $step
+    }
+
+    [IArgumentCompleter] Create() { return [NumberCompleter]::new($this.From, $this.To, $this.Step) }
+}
+```
+
+Usage from PowerShell would then be:
+
+```powershell
+function Add{
+    param(
+       [NumberCompletions(0, 100, 5)]
+       [int] $X
+      ,
```

Comment:
should put the comma at end of preceding line

Reply:
Weird, don't know what happened there. Will fix.